### PR TITLE
Added Elasticsearch index

### LINKED ISSUE #670: search: re-index when user profile is created / updated

### LINKED ISSUE #699: Implement ElasticSearch index
we'll use ElasticSearch DSL

Configure the index schema to match the user profile, user education, user employment

We will be adding new fields to the models in other issues. #677 #672

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #699 
Fixes #670 

#### What's this PR do?
Sets up Elasticsearch indexing to happen:
 - when the `recreate_index` management command is run.
 - when a `Profile`, `Education`, or `Employment` are updated or created. In this case a celery task is executed

#### Where should the reviewer start?
search/api.py

#### How should this be manually tested?
There are changes to `docker-compose.yml` so you'll need to restart your containers, no rebuilding should be necessary though.

First, run `docker-compose run web ./manage.py recreate_index`. Then run this command: `docker-compose run web curl http://elastic:9200/micromasters/_search | python -m json.tool`. Verify that you see profile data in the search results.

Go into the users page and make an update to the profile. Run the curl command again and verify that you see the updated value in the profile.

#### Background context
From now on, people creating new instances of this app will need to create an Elasticsearch index and mapping via management command: `./manage.py recreate_index`. This command may also be run at any time to delete the existing index and reindex all relevant data (this may take a while).

### CODE REVIEW

**Comment by giocalitri** [comment]
@gsidebo Feel free to review, but I would like to make a pass at the very end to see if I spot something odd

**Comment by aliceriot** [comment]
I ran into an issue with CORS headers when trying to hit the ES index directly from the JS client:

cors

We probably will solve this issue when we build out the search API.

**Comment by gsidebo** [comment]
> From now on, people creating new instances of this app will need to create an Elasticsearch index and mapping via management command:

does this need to be added to the readme?

File: search/base.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+"""
```
Comment (gsidebo) [inline search/base.py]: i would not expect to find our base test case class for ElasticSearch-related functionality, or any test case class for that matter, in `search/base.py`. i don't know why it's in the search/ directory, and base.py is misleading. out of the directories we have, i would think a new file (`test_classes.py`, et al) in `micromasters/tests` would make more sense, but maybe we want a new directory

File: search/base.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+Base test classes for search
+"""
+
+from django.test import TestCase
+
```
Comment (gsidebo) [inline search/base.py]: any reason for the empty line?

File: search/apps.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+"""
+AppConfig
+"""
+from django.apps import AppConfig
```
Comment (gsidebo) [inline search/apps.py]: let's be consistent about whether or not we're inserting an empty line before the imports

File: search/api_test.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+class SerializerTests(ESTestCase):
+    """
+    Tests for serializer
+    """
+
+    def test_profile_serializer(self):  # pylint: disable=no-self-use
+        """
+        Tests the profile serializer
+        """
```
Comment (gsidebo) [inline search/api_test.py]: are docstrings a hard requirement for the linter? if so, this is an example of when that kinda hurts us. this docstring is really just taking up space and doesn't add any information. not that it _needs_ more information - `test_profile_serializer` describes itself just fine

File: search/api_test.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        # Confirm nothing in index
+        assert_search(search(), [])
+        index_profiles(Profile.objects.iterator(), chunk_size=4)
+        assert_search(search(), list(Profile.objects.all()))
+
+
+class SerializerTests(ESTestCase):
+    """
+    Tests for serializer
+    """
```
Comment (gsidebo) [inline search/api_test.py]: ditto for the comment below?

File: search/api_test.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    def test_profile_update(self):
+        """
+        Test that Profile is reindexed after being updated
+        """
+        user = UserFactory.create()
+        assert search()['total'] == 1
+        profile = user.profile
+        profile.first_name = 'updated'
+        profile.save()
+
```
Comment (gsidebo) [inline search/api_test.py]: can we get rid of these empty lines before the assertion?

File: search/api.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+log = logging.getLogger(__name__)
+
+
+PROFILE_DOC_TYPE = 'profile'
+
+DOC_TYPES = (PROFILE_DOC_TYPE, )
+
+_CONN = None
+# When we create the connection, check to make sure all appropriate mappings exist
+_CONN_VERIFIED = False
```
Comment (gsidebo) [inline search/api.py]: can we get rid of the blank lines from line 20 to here?

File: search/base.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+class ESTestCase(TestCase):
+    """
+    Set up ES index on setup
+    """
+
+    def setUp(self):
+        super(ESTestCase, self).setUp()
+
+        clear_index()
```
Comment (gsidebo) [inline search/base.py]: empty line

File: search/api.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    """
+    Wipe and recreate index and mapping. No indexing is done.
+    """
+    conn = get_conn(verify=False)
+    index_name = settings.ELASTICSEARCH_INDEX
+    if conn.indices.exists(index_name):
+        conn.indices.delete(index_name)
+    conn.indices.create(index_name)
+    conn.indices.refresh()
+
```
Comment (gsidebo) [inline search/api.py]: sorry to harp on the empty lines across these files, but this one and line 206 don't seem to add any value

File: search/api.py (hunk 7 above)
Comment (gsidebo) [inline search/api.py]: i'm ok with the usage of `global` in the function(s) below, but i think it would help to encapsulate `_CONN` and `_CONN_VERIFIED` in a simple class, eg: `ESConnection`?  lines 24 & 26 would become something like `es_connection = ESConnection()`. some other possible changes as a result:

- line 38: `if _CONN is None` --> `if not es_connection.created():`
- line 39: `_CONN = connections.create_connection...` --> `es_connection.set(connections.create_connection...)`
- line 43: `if verify and not _CONN_VERIFIED:` --> `if verify and not es_connection.verified():`
- create helper methods for `_CONN.indices.exists...`, `_CONN.indices.get_mapping()...` etc

since `get_conn` is used in a bunch of places, maybe it should still return that object (ie: `return es_connection.conn`)

**Comment by gsidebo** [comment]
functionality looks good

**Comment by noisecapella** [comment]
I was just thinking that this requires #707 before it gets merged

File: search/base.py (hunk 1 above)
Comment (noisecapella) [inline search/base.py]: The `_test.py` files are supposed to be located next to the files they test so that they show up next to each other in code review. I'm not sure it makes sense to move `search/base.py` into a `tests` directory by itself

File: search/api_test.py (hunk 4 above)
Comment (noisecapella) [inline search/api_test.py]: I still think it's a good idea for them to be mandatory, though this function is small enough to not need it. I'll reword this docstring to describe the test better

File: search/api.py (hunk 7 above)
Comment (noisecapella) [inline search/api.py]: I'm not sure this gives us much. We'll still need some global to store the `ESConnection`, and the helper function `verify` would only be called in one place. `get_conn` already encapsulates the verification so that it happens when a connection is requested

**Comment by noisecapella** [comment]
Per my conversation with @giocalitri I'm also going to change the indexing to be on the user rather than profile

File: search/base.py (hunk 1 above)
Comment (gsidebo) [inline search/base.py]: there are many suggestions for structuring a python project and placement of unit tests, and [a](http://docs.python-guide.org/en/latest/writing/structure/#sample-repository) [few](http://learnpythonthehardway.org/book/ex46.html) of them contradict the structure of this codebase. i don't really agree with using code review convenience as a guide to structuring a project, and i can't say that i've ever found it really beneficial to have the two files next to each other, though others' results may vary. on top of that, if each project folder had a tests/ directory (where necessary), the file with the actual logic and the test file would be reasonably proximal anyway.

given the structure of the project elsewhere, i'm ok with leaving this as-is, but i think this is a pretty good illustration of the downfalls of mixing our test modules in with 'real' modules. we're putting a test class into a module whose name does not suggest that it has anything to do with tests. we might have other functionality (perhaps the ESConnection class i suggested elsewhere in the PR) that makes sense in `search/base.py`. then we have 'real' code mixed in with test 'code'. not the end of the world, but i think we should be avoiding that

**Comment by noisecapella** [comment]
Ready for review again. I also added an explicit ES mapping which I forgot in the earlier revision

@giocalitri Can you take a look when you have time?

File: search/api.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+from search.exceptions import ReindexException
+
+log = logging.getLogger(__name__)
+
+
+USER_DOC_TYPE = 'user'
+DOC_TYPES = (USER_DOC_TYPE, )
+_CONN = None
+# When we create the connection, check to make sure all appropriate mappings exist
+_CONN_VERIFIED = False
```
Comment (gsidebo) [inline search/api.py]: any opinion on the change to `_CONN` and `_CONN_VERIFIED` that i mentioned earlier?

Comment (noisecapella) [inline search/api.py]: I don't think it's worth encapsulating into `ESConnection` since we still need one global variable to store the `ESConnection`. Both variables are only used by `get_conn`

File: search/api.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        }},
+    })
+    mapping.save(index_name)
+
+
+def create_mappings():
+    """
+    Create all mappings, deleting existing mappings if they exist.
+    """
+    create_user_mapping()
```
Comment (gsidebo) [inline search/api.py]: do we have a reason for both `create_mappings` and `create_user_mapping` to exist?

Comment (noisecapella) [inline search/api.py]: It's there in case we want to add other mappings in the future

File: search/api.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        conn.indices.delete_mapping(index=index_name, doc_type=USER_DOC_TYPE)
+
+    # A string which must be indexed exactly as is and not stemmed for full text search (the default)
+    not_analyzed_string_type = {
+        'type': 'string',
+        'index': 'not_analyzed',
+    }
+    bool_type = {'type': 'boolean'}
+    date_type = {'type': 'date', 'format': 'date'}
+    long_type = {'type': 'long'}
```
Comment (gsidebo) [inline search/api.py]: if we're setting `USER_DOC_TYPE` as a module-level variable, i think it makes sense to do the same for the above 4 values

File: search/apps.py (hunk 3 above)
Comment (gsidebo) [inline search/apps.py]: fair enough, im not too insistent on that one

File: search/api.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+from elasticsearch.helpers import bulk
+from elasticsearch.exceptions import NotFoundError
+from elasticsearch_dsl import Mapping
+from elasticsearch_dsl.connections import connections
+
+from profiles.models import Profile
+from profiles.serializers import ProfileSerializer
+from search.exceptions import ReindexException
+
+log = logging.getLogger(__name__)
```
Comment (gsidebo) [inline search/api.py]: since this isn't an import, shouldn't this line have 2 blank lines preceding it and, if preferred, just 1 blank line succeeding it?

**Comment by gsidebo** [comment]
looks good aside from the two small comments. testing functionality again

File: search/api.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
+"""
+Functions for search
+"""
+from itertools import islice
+import logging
```
Comment (giocalitri) [inline search/api.py]: `import ` imports come before `from` imports

File: search/api.py (hunk 13 above)
Comment (giocalitri) [inline search/api.py]: nope: Class and function definition have to

File: search/api.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+
+USER_DOC_TYPE = 'user'
+DOC_TYPES = (USER_DOC_TYPE, )
+_CONN = None
+# When we create the connection, check to make sure all appropriate mappings exist
+_CONN_VERIFIED = False
+
+
+def get_conn(verify=True):
```
Comment (giocalitri) [inline search/api.py]: this does come directly from LORE, right?

File: search/api.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+            List of users
+
+    Returns:
+        int: Number of items inserted into Elasticsearch
+    """
+
+    conn = get_conn()
+    insert_count, errors = bulk(
+        conn,
+        (serialize_user(user) for user in users),
```
Comment (giocalitri) [inline search/api.py]: nit: why did you put parentheses here?

File: search/api.py (hunk 13 above)
Comment (gsidebo) [inline search/api.py]: then i would suggest we get rid of the double line break under it

File: search/api.py (hunk 15 above)
Comment (noisecapella) [inline search/api.py]: In general yes. Here's the LORE code if you want to compare: https://github.com/mitodl/lore/blob/master/search/utils.py#L49

File: search/api.py (hunk 16 above)
Comment (noisecapella) [inline search/api.py]: In generators parentheses are required: https://docs.python.org/3/reference/expressions.html#generator-expressions

File: search/api.py (hunk 14 above)
Comment (noisecapella) [inline search/api.py]: Is this our convention or a PEP 8 requirement? I'm happy to make the change, just haven't heard of this before

File: search/api.py (hunk 10 above)
Comment (gsidebo) [inline search/api.py]: `get_conn` could be made shorter and more readable if we created a class that took care of some of these concerns, but i don't feel too insistent on this one. carry on

File: search/api.py (hunk 14 above)
Comment (giocalitri) [inline search/api.py]: More like a general convention 
http://stackoverflow.com/a/20763446/257092
But feel free to leave it as is

File: courses/views_tests.py [hunk 17]
```diff
@@ -1,12 +1,12 @@
...
 # pylint: disable=no-self-use
 from django.core.urlresolvers import reverse
-from django.test import TestCase
 
 from .factories import ProgramFactory, CourseFactory
+from search.base import ESTestCase
 
 
-class ProgramTests(TestCase):
+class ProgramTests(ESTestCase):
```
Comment (giocalitri) [inline courses/views_tests.py]: why this change?

Comment (giocalitri) [inline courses/views_tests.py]: wait, I have still to read the new mixin

Comment (giocalitri) [inline courses/views_tests.py]: ok... I remember now

Comment (noisecapella) [inline courses/views_tests.py]: We need to run `recreate_index()` before each test to allow signals sent when the models change to connect to Elasticsearch correctly

File: search/api_test.py [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+def assert_search(results, users):
+    """
+    Assert that search results match users
+    """
+    assert results['total'] == len(users)
+    sources = sorted([hit['_source'] for hit in results['hits']], key=lambda hit: hit['id'])
+    sorted_users = sorted(users, key=lambda user: user.id)
+    serialized = [remove_key(serialize_user(user), "_id") for user in sorted_users]
+    assert serialized == sources
```
Comment (giocalitri) [inline search/api_test.py]: I would put these two functions in a mixin, but feel free to leave as is

File: search/base.py [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class ESTestCase(TestCase):
+    """
+    Set up ES index on setup
+    """
+
+    def setUp(self):
+        super(ESTestCase, self).setUp()
+        clear_index()
```
Comment (giocalitri) [inline search/base.py]: we might want to add an extra function to disable the indexing while we perform some actions in a test: I would prefer to avoid the indexing at all in some cases where I do not need it.
But feel free to ignore for now (and maybe add another issue)

**Comment by giocalitri** [comment]
I think this code is fine

File: search/base.py (hunk 19 above)
Comment (noisecapella) [inline search/base.py]: Filed #721 for this

**Comment by noisecapella** [comment]
Comments addressed, any other comments @gsidebo?

**Comment by gsidebo** [comment]
looks great 👍
